File: src/main/java/nz/co/searchwellington/controllers/models/TagGeotaggedModelBuilder.java
package nz.co.searchwellington.controllers.models;

import java.util.List;

import javax.servlet.http.HttpServletRequest;

import nz.co.searchwellington.controllers.RssUrlBuilder;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.model.frontend.FrontendResource;
import nz.co.searchwellington.repositories.ContentRetrievalService;
import nz.co.searchwellington.urls.UrlBuilder;

import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.ModelAndView;

@Component
public class TagGeotaggedModelBuilder extends AbstractModelBuilder implements ModelBuilder {
	
	private static Logger log = Logger.getLogger(TagGeotaggedModelBuilder.class);
    	
	private ContentRetrievalService contentRetrievalService;
	private UrlBuilder urlBuilder;
	private RssUrlBuilder rssUrlBuilder;

	@Autowired
	public TagGeotaggedModelBuilder(
			ContentRetrievalService contentRetrievalService,
			UrlBuilder urlBuilder, RssUrlBuilder rssUrlBuilder) {
		this.contentRetrievalService = contentRetrievalService;
		this.urlBuilder = urlBuilder;
		this.rssUrlBuilder = rssUrlBuilder;
	}

	
	@Override
	@SuppressWarnings("unchecked")
	public boolean isValid(HttpServletRequest request) {
		List<Tag> tags = (List<Tag>) request.getAttribute("tags");
		boolean isSingleTagPage = tags != null && tags.size() == 1;
		boolean hasCommentPath = request.getPathInfo().matches("^(.*?)/geotagged(/(rss|json))?$");		
		return isSingleTagPage && hasCommentPath;
	}
	
	@Override
	@SuppressWarnings("unchecked")
	public ModelAndView populateContentModel(HttpServletRequest request) {
		if (isValid(request)) {
			log.debug("Building tag geotagged page model");
			List<Tag> tags = (List<Tag>) request.getAttribute("tags");
			Tag tag = tags.get(0);
			return populateTagCommentPageModelAndView(tag);
		}
		return null;
	}
	
	@Override
	public void populateExtraModelContent(HttpServletRequest request, ModelAndView mv) {	
	}
	
	@Override
	public String getViewName(ModelAndView mv) {
		return "tagGeotagged";
	}
	
	private ModelAndView populateTagCommentPageModelAndView(Tag tag) {		
		ModelAndView mv = new ModelAndView();				
		mv.addObject("tag", tag);
		mv.addObject("heading", tag.getDisplayName() + " geotagged");        		
		mv.addObject("description", "Geotagged " + tag.getDisplayName() + " newsitems");
		mv.addObject("link", urlBuilder.getTagCommentUrl(tag));		
	    final List<FrontendResource> allGeotaggedForTag = contentRetrievalService.getTaggedGeotaggedNewsitems(tag, MAX_NUMBER_OF_GEOTAGGED_TO_SHOW);
		mv.addObject("main_content", allGeotaggedForTag);
		if (allGeotaggedForTag.size() > 0) {
			 setRss(mv, rssUrlBuilder.getRssTitleForTagGeotagged(tag), rssUrlBuilder.getRssUrlForTagGeotagged(tag));
		}		
		return mv;
	}
			
}

File: src/main/scala/nz/co/searchwellington/controllers/models/helpers/TagGeotaggedModelBuilder.scala
package nz.co.searchwellington.controllers.models.helpers

import io.opentelemetry.api.trace.Span
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.controllers.RssUrlBuilder
import nz.co.searchwellington.filters.RequestPath
import nz.co.searchwellington.filters.attributesetters.TagPageAttributeSetter
import nz.co.searchwellington.model.{Tag, User}
import nz.co.searchwellington.repositories.ContentRetrievalService
import nz.co.searchwellington.urls.UrlBuilder
import org.apache.commons.logging.LogFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.ui.ModelMap

import javax.servlet.http.HttpServletRequest
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._

@Component class TagGeotaggedModelBuilder @Autowired()(val contentRetrievalService: ContentRetrievalService,
                                                       urlBuilder: UrlBuilder, rssUrlBuilder: RssUrlBuilder,
                                                       commonAttributesModelBuilder: CommonAttributesModelBuilder) extends ModelBuilder
  with CommonSizes with ReasonableWaits {

  private val log = LogFactory.getLog(classOf[TagGeotaggedModelBuilder])

  def isValid(request: HttpServletRequest): Boolean = {
    val tags = request.getAttribute(TagPageAttributeSetter.TAGS).asInstanceOf[Seq[Tag]]
    val isSingleTagPage = tags != null && tags.size == 1
    val hasGeotaggedSuffix = RequestPath.getPathFrom(request).matches("^(.*?)/geotagged(/(rss|json))?$")
    isSingleTagPage && hasGeotaggedSuffix
  }

  def populateContentModel(request: HttpServletRequest, loggedInUser: Option[User])(implicit ec: ExecutionContext, currentSpan: Span): Future[Option[ModelMap]] = {
    def populateTagCommentPageModelAndView(tag: Tag): Future[Some[ModelMap]] = {
      for {
        newsitems <- contentRetrievalService.getGeotaggedNewsitemsForTag(tag, MAX_NUMBER_OF_GEOTAGGED_TO_SHOW, loggedInUser = loggedInUser)
      } yield {
        val mv = new ModelMap().
          addAttribute("tag", tag).
          addAttribute("heading", tag.getDisplayName + " geotagged").
          addAttribute("description", "Geotagged " + tag.getDisplayName + " newsitems").
          addAttribute("link", urlBuilder.fullyQualified(urlBuilder.getTagGeocodedUrl(tag))).
          addAttribute(MAIN_CONTENT, newsitems.asJava)

        if (newsitems.nonEmpty) {
          commonAttributesModelBuilder.setRss(mv, rssUrlBuilder.getRssTitleForTagGeotagged(tag), rssUrlBuilder.getRssUrlForTagGeotagged(tag))
        }
        Some(mv)
      }
    }

    log.debug("Building tag geotagged page model")
    val tags = request.getAttribute(TagPageAttributeSetter.TAGS).asInstanceOf[Seq[Tag]]
    val tag = tags.head
    populateTagCommentPageModelAndView(tag)
  }

  def populateExtraModelContent(request: HttpServletRequest, loggedInUser: Option[User])(implicit ec: ExecutionContext, currentSpan: Span): Future[ModelMap] = {
    latestNewsitems(loggedInUser)
  }

  def getViewName(mv: ModelMap, loggedInUser: Option[User]): String = "tagGeotagged"

}
